A80-grpc-metrics-for-tcp-connection

```diff
@@ -0,0 +1,62 @@
+A80: gRPC Metrics for TCP connection
+----
+* Author(s):  Yash Tibrewal (@yashykt), Nana Pang (@nanahpang), Yousuk Seung (@yousukseung)
+* Approver: Craig Tiller (@ctiller), Mark Roth (@markdroth)
+* Status: {Draft, In Review, Ready for Implementation, Implemented}
+* language: {...}
+* Last updated: 2024-04-18
+* Discussion at: {...}
+
+## Abstract
+
+This document proposes adding new TCP connection metrics to gRPC for improved network analysis and debugging.
+
+## Background
+
+To improve the network debugging capabilities for gRPC users, we propose adding per-connection TCP metrics in gRPC. The metrics will utilize the metrics framework outlined in  [A79].
+
+### Related Proposals: 
+* [A79]: gRPC Non-Per-Call Metrics Framework (pending)
+
+[A79]: https://github.com/grpc/proposal/pull/421
+
+## Proposal
+
+This document proposes changes to the following gRPC components.
+
+#### Per-Connection TCP Metrics
+
+We will provide the following metrics:
+- `grpc.tcp.min_rtt`
+- `grpc.tcp.delivery_rate`
+- `grpc.tcp.packets_sent`
+- `grpc.tcp.packets_retransmitted`
+- `grpc.tcp.packets_spurious_retransmitted`
+
+The metrics will have label:
+
+| Name        | Disposition | Description |
+| ----------- | ----------- | ----------- |
+| grpc.tcp.remote_peer_address | optional | Store the peer address info in the format as `ip:port`. |
```

Comment:
@markdroth @yashykt I feel like there's an equivalent label already in use?

Reply:
This could be a relatively expensive fan-out for a busy server

Reply:
Agreed, we discussed about this offline. Is there a way to support an opaque user-defined peer info string? In prod, by default we only record the cell information due to the fan-out concern.

Reply:
Not yet

Reply:
I propose URI form, such as `ipv4:1.2.3.4:567`.  @ejona86 @dfawley, would that be a problem for Java and Go if we ever added these metrics in your languages?

Reply:
I think this should be clarified more.  For instance, is `ipv4:///1.2.3.4:567` also OK?  What is the format for `ipv6`?  What other schemes do we support?  @ejona86 is concerned that `ipv4:1.2.3.4:567` is not a URI we should support in the first place.  So maybe this should just be specified fully as strictly:

```
- ipv4:<ipv4-address>:<port-number>
OR
- ipv6:[<ipv6-address>]:<port-number>
```